src/native/eventpipe/ep.c

```diff
@@ -571,11 +571,84 @@ disable_helper (EventPipeSessionID id)
 		EventPipeProviderCallbackDataQueue callback_data_queue;
 		EventPipeProviderCallbackData provider_callback_data;
 		EventPipeProviderCallbackDataQueue *provider_callback_data_queue = ep_provider_callback_data_queue_init (&callback_data_queue);
+		EventPipeSession *session = NULL;
+		EventPipeThread * thread = NULL;
 
-		EP_LOCK_ENTER (section1)
-			if (ep_volatile_load_number_of_sessions () > 0)
-				disable_holding_lock (id, provider_callback_data_queue);
-		EP_LOCK_EXIT (section1)
+		if (ep_volatile_load_number_of_sessions () > 0) {
+			// Do setup of disable under lock
+			EP_LOCK_ENTER (section0)
+				EP_ASSERT (id != 0);
+				EP_ASSERT (ep_volatile_load_number_of_sessions () > 0);
+
+				if (is_session_id_in_collection (id)) { // requires lock
+					session = (EventPipeSession *)id;
+
+					if (session_requested_sampling (session)) {
+						// Disable the profiler.
+						ep_sample_profiler_disable (); // requires lock
+					}
+
+					// Log the process information event.
+					log_process_info_event (ep_event_source_get ());
+
+					// Disable session tracing.
+					config_enable_disable (ep_config_get (), session, provider_callback_data_queue, false); // Requires lock
+
+					ep_session_disable (session); // WriteAllBuffersToFile, and remove providers.
+
+					if (ep_session_get_rundown_requested (session) && _ep_can_start_threads) {
+						ep_session_enable_rundown (session); // Set Rundown provider.
+						thread = ep_thread_get_or_create ();
+						if (thread != NULL) {
+							ep_thread_set_as_rundown_thread (thread, session);
+							config_enable_disable (ep_config_get (), session, provider_callback_data_queue, true);
+						} else {
+							EP_ASSERT (!"Failed to get or create the EventPipeThread for rundown events.");
+						}
+					}
+				}
+			EP_LOCK_EXIT (section0)
+
+			// Do rundown
+			if (session != NULL) {
+				if (thread != NULL && ep_session_get_rundown_requested (session) && _ep_can_start_threads) {
+					ep_session_execute_rundown (session);
+					ep_thread_set_as_rundown_thread (thread, NULL);
```

Comment:
What happens if you try to terminate the session a second time during rundown? Before we held the lock so the second request wouldn't execute until after rundown was complete and everything was shut down. Now we could theoretically have someone issue a second disable while rundown is happening.

Reply:
The main mechanism we have for issuing Disable commands is the IPC server, which currently is single-threaded. While not ideal because it is implicit, we have an implicit serialization of IPC commands.

That being said, I don't think there is anything that is wrong about having two rundowns happening at the same time. You may end up with duplicate information in both traces, but the runtime shouldn't run into any issues. The reader might need to be updated to handle duplicates if it doesn't already.

The things that should be exclusive are still wrapped in the spin lock, e.g., EP config changes, non-volatile state updates.